vllm/block.py

```diff
@@ -23,7 +47,13 @@ def __init__(
         self.block_number = block_number
         self.block_size = block_size
 
-        self.token_ids = [_BLANK_TOKEN_ID] * block_size
+        self.token_ids = _BLOCK_POOL.alloc_block(block_size)
+        # this finalizer is used to return the block to the pool when the object is deleted # noqa
+        # NOTE: don't use __del__ because it cannot guarantee the order of finalization, # noqa
+        # i.e. `self.token_ids` may be deleted before `self`, and we lose
+        #  the opportunity to return the block to the pool
+        self._finalizer = weakref.finalize(self, _BLOCK_POOL.del_block,
```

Comment:
Although not as "automatic", it may be more efficient to call a free method explicitly when a `LogicalTokenBlock` is finished with. Since like you say here there is no guarantee about when finalization happens. In my experience it's better to avoid relying on finalization wherever possible.

Reply:
> it may be more efficient to call a free method explicitly when a LogicalTokenBlock is finished with

adding a `LogicalTokenBlock.release()` method is easy, but it would be difficult to find all the places to call it. Essentially we need to mimic and track the gc information. Given the complexity, I prefer the `finalization` method, this will guarantee that `self` is destructed before `self.token_ids`

Reply:
I haven't checked but maybe there aren't that many places? In any case it doesn't matter if some place is missed. Either it can be dropped in that case or keep the existing logic to let the finalizer get it...

Reply:
I don't have the bandwidth to check it yet. Feel free to add it later if you figure it out. My intuition is this would need control over the gc system (which is difficult in Python).